Source/AdUnits/Native/NativeAd.swift

```diff
@@ -0,0 +1,320 @@
+        && canOpenString(iconUrl)
+        && canOpenString(imageUrl)
+        && canOpenString(clickUrl)
+    }
+    
+    private func canOpenString(_ string: String?) -> Bool {
+        guard let string = string else {
+            return false
+        }
+        let detector = try! NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)
+        if let match = detector.firstMatch(in: string, options: [], range: NSRange(location: 0, length: string.utf16.count)) {
+            return match.range.length == string.utf16.count
+        } else {
+            return false
+        }
+    }
+    
+    //MARK: registerView function
+    @discardableResult 
+    public func registerView(view :UIView, clickableViews :[UIView]? ) -> Bool{
+        guard view != nil else {
+            Log.error("A valid view is required for tracking")
+            return false
+        }
+        guard !expired else {
+            Log.error("The native Ad is expired, cannot use it for tracking")
+            return false
+        }
+        viewForTracking = view
+        setupViewabilityTracker()
+        attachGestureRecognizersToNativeView(nativeView: view, withClickableViews: clickableViews)
+        return true
+    }
+    
+    private func unregisterViewFromTracking(){
+        detachAllGestureRecognizers()
+        viewForTracking = nil
+        invalidateTimer(viewabilityTimer)
+    }
+    
+    //MARK: NativeAd Expire
+    func cacheExpired() {
+        if viewForTracking == nil {
+            expired = true
+            delegate?.adDidExpire?(ad: self)
+            unregisterViewFromTracking()
+        }
+    }
+    
+    private func invalidateTimer(_ timer :Timer?){
+        if let timer = timer, timer.isValid {
+            timer.invalidate()
+        }
+    }
+    
+    
+    //MARK: Impression Tracking
+    private func setupViewabilityTracker(){
+        let requiredAmountOfSimultaneousViewableEvents = lround(NativeAd.kNativeAdIABShouldBeViewableForTrackingDuration / NativeAd.kNativeAdCheckViewabilityForTrackingFrequency) + 1
+        
+        targetViewabilityValue = lround(pow(Double(2),Double(requiredAmountOfSimultaneousViewableEvents)) - 1)
+        
+        Log.debug("\n\trequiredAmountOfSimultaneousViewableEvents=\(requiredAmountOfSimultaneousViewableEvents) \n\ttargetViewabilityValue=\(targetViewabilityValue)")
+
+        viewabilityTimer = Timer.scheduledTimer(withTimeInterval: NativeAd.kNativeAdCheckViewabilityForTrackingFrequency, repeats: true) { [weak self] timer in
+            guard let strongSelf = self else {
+                timer.invalidate()
+                Log.debug("FAILED TO ACQUIRE strongSelf viewabilityTimer")
+                return
+            }
+            strongSelf.checkViewability()
+        }
+    }
+    
+    @objc private func checkViewability() {
+        viewabilityValue = (viewabilityValue << 1 | (viewForTracking?.an_isAtLeastHalfViewable() == true ? 1 : 0)) & targetViewabilityValue
```

Comment:
can you  please explain these bit operations ?

Reply:
1. Initially, viewabilityValue value will be 0, and checkViewability() method will keep calling unless targetViewabilityValue will be equal to viewabilityValue
2. Inside checkViewability() method, with the help of this method `an_isAtLeastHalfViewable()` we constantly monitor view viewability at least half viewable on the screen and once the view is half viewable, then returns true which further returns 1 with the help of a ternary operator
3. Then viewabilityValue will be calculated based on bitwise left shift operator (<<) to 1 and compares bits of these two values with bitwise OR operator (|) which further compares bits of viewabilityValue and targetViewabilityValue with bitwise AND operator (&)
4. And once both values are equal then we fire the impressions.

Reply:
@anwzhang, do we use the same logic of Impression Tracking for Android ?

Reply:
@yoalex5 we use similar logic, we have a visibility detector that monitors the viewabibilty of a registered view.